lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,139 @@
+
+Description:
+List outstanding certificates. If --all is specified, signed certificates will be listed as well.
+
+Options:
+      BANNER
+
+      BODY = JSON.dump({desired_state: 'signed'})
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--config PUPPET.CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'List signed certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        requested, signed = separate_certs(all_certs)
+        if input['all']
+          human_readable(requested, signed)
+        else
+          human_readable(requested)
+        end
+        return 0
+      end
+
+      def human_readable(requested, signed = nil)
+        if signed && signed.size != 0
+          @logger.inform "Signed Certificates:"
+          print_output(signed)
+        end
+
+        if requested.size != 0
+          @logger.inform "Requested Certificates:"
+          print_output(requested)
+        end
+      end
+
+      def print_output(certs)
+        padded = 0
+        certs.each do |cert|
+          cert_size = cert["name"].size
+          padded =  cert_size if cert_size > padded
+        end
+
+        certs.each do |cert|
+          @logger.inform "    #{cert["name"]}".ljust(padded + 6) + " (SHA256) #{cert["fingerprints"]["SHA256"]}" +
+                             (cert["dns_alt_names"].empty? ? "" : "\talt names: #{cert["dns_alt_names"]}")
+          end
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def separate_certs(all_certs)
+        requested = all_certs.find_all { |v| v["state"] == "requested" }
+        signed    = all_certs.find_all { |v| v["state"] == "signed" }
+        return requested, signed
+      end
+
+      def get_all_certs(settings)
+        result = get_certificate_statuses(settings)
+
+        unless result.is_a? Net::HTTPOK
```

Comment:
So if there is an error, we will want to return a 1 from the run method.

Reply:
Also, maybe we want to specifically catch the NO CONTENT.  So that its not really an error, there just aren't any certificates.  I'm assuming no certificates returns 204?  or?